services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DownloadDirectoryHelper.java

```diff
@@ -0,0 +1,225 @@
+                                                      + "directory", directory);
+    }
+
+    private void doDownloadDirectory(CompletableFuture<CompletedDirectoryDownload> returnFuture,
+                                     DownloadDirectoryRequest downloadDirectoryRequest) {
+        validateDirectory(downloadDirectoryRequest);
+        String bucket = downloadDirectoryRequest.bucket();
+        String delimiter = getDelimiter(downloadDirectoryRequest);
+        String prefix = downloadDirectoryRequest.prefix().orElse(DEFAULT_PREFIX);
+
+        ListObjectsV2Request request = ListObjectsV2Request.builder()
+                                                           .bucket(bucket)
+                                                           .prefix(prefix)
+                                                           .delimiter(delimiter)
+                                                           .build();
+
+        Collection<FailedFileDownload> failedFileDownloads = new ConcurrentLinkedQueue<>();
+        List<CompletableFuture<?>> futures = new ArrayList<>();
+
+        listObjectsRecursivelyHelper.s3Objects(request).subscribe(s3Object -> {
+            log.debug(() -> "s3Object key: " + s3Object.key());
+            futures.add(downloadSingleFile(downloadDirectoryRequest,
+                                           failedFileDownloads,
+                                           s3Object));
+        }).whenComplete((r, t) -> {
+            if (t != null) {
+                returnFuture.completeExceptionally(SdkClientException.create("Failed to call ListObjectsV2  ", t));
+            } else {
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+                                 .whenComplete((response, throwable) -> returnFuture.complete(
+                                     CompletedDirectoryDownload.builder()
+                                                               .failedTransfers(failedFileDownloads)
+                                                               .build()));
+            }
+        });
+    }
+
+    private static String getDelimiter(DownloadDirectoryRequest downloadDirectoryRequest) {
+        return downloadDirectoryRequest.delimiter().orElse(DEFAULT_DELIMITER);
+    }
+
+    private CompletableFuture<CompletedFileDownload> downloadSingleFile(DownloadDirectoryRequest downloadDirectoryRequest,
+                                                                        Collection<FailedFileDownload> failedFileDownloads,
+                                                                        S3Object s3Object) {
+        String directory = downloadDirectoryRequest.destinationDirectory().toString();
+        String relativePath = getRelativePath(getDelimiter(downloadDirectoryRequest),
+                                              s3Object.key());
+
+        Path destinationPath = fileSystem.getPath(directory, relativePath);
```

Comment:
I think this will potentially be non-intuitive behavior.  `DownloadDirectoryRequest.destinationDirectory()` is a `Path`, and those are tied to a specific filesystem (they have `Path#getFileSystem()`). So if their `destinationDirectory` was created using something like `myCustomFs.getPath(..)`, I'd expect the file to end up in `myCustomFs`, instead of the system default filesystem.

I think in all cases, anywhere we need to refer to the filesystem, it should be through `destinationDirectory.getFileSystem()`. I think that also means you don't need to take a `FileSystem` parameter and can handle any number of them.